File: android/app/src/main/java/my/ssdid/wallet/domain/mdoc/MDoc.kt
package my.ssdid.wallet.domain.mdoc

data class StoredMDoc(
    val id: String,
    val docType: String,
    val issuerSignedCbor: ByteArray,
    val deviceKeyId: String,
    val issuedAt: Long,
    val expiresAt: Long? = null,
    val nameSpaces: Map<String, List<String>> = emptyMap()
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is StoredMDoc) return false
        return id == other.id
    }

    override fun hashCode() = id.hashCode()
}

data class IssuerSigned(
    val nameSpaces: Map<String, List<IssuerSignedItem>>,
    val issuerAuth: ByteArray
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is IssuerSigned) return false
        return nameSpaces == other.nameSpaces && issuerAuth.contentEquals(other.issuerAuth)
    }

    override fun hashCode() = nameSpaces.hashCode()
}

data class IssuerSignedItem(
    val digestId: Int,
    val random: ByteArray,
    val elementIdentifier: String,
    val elementValue: Any
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is IssuerSignedItem) return false
        return digestId == other.digestId && elementIdentifier == other.elementIdentifier
    }

    override fun hashCode() = 31 * digestId + elementIdentifier.hashCode()
}

data class MobileSecurityObject(
    val version: String,
    val digestAlgorithm: String,
    val valueDigests: Map<String, Map<Int, ByteArray>>,
    val deviceKeyInfo: DeviceKeyInfo,
    val validityInfo: ValidityInfo
)

data class DeviceKeyInfo(
    val deviceKey: ByteArray
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is DeviceKeyInfo) return false
        return deviceKey.contentEquals(other.deviceKey)
    }

    override fun hashCode() = deviceKey.contentHashCode()
}

data class ValidityInfo(
    val signed: String,
    val validFrom: String,
    val validUntil: String
)

File: ios/SsdidWallet/Domain/MDoc/MDoc.swift
import Foundation

/// Persistent metadata for a stored mdoc / mDL credential.
struct StoredMDoc: Codable, Identifiable, Equatable {
    let id: String
    let docType: String
    let issuerSignedCbor: Data
    let deviceKeyId: String
    let issuedAt: Int64
    let expiresAt: Int64?
    let nameSpaces: [String: [String]]

    init(
        id: String,
        docType: String,
        issuerSignedCbor: Data,
        deviceKeyId: String,
        issuedAt: Int64,
        expiresAt: Int64? = nil,
        nameSpaces: [String: [String]] = [:]
    ) {
        self.id = id
        self.docType = docType
        self.issuerSignedCbor = issuerSignedCbor
        self.deviceKeyId = deviceKeyId
        self.issuedAt = issuedAt
        self.expiresAt = expiresAt
        self.nameSpaces = nameSpaces
    }

    static func == (lhs: StoredMDoc, rhs: StoredMDoc) -> Bool {
        lhs.id == rhs.id
    }
}

/// Decoded IssuerSigned structure from ISO 18013-5.
struct IssuerSigned: Equatable {
    let nameSpaces: [String: [IssuerSignedItem]]
    let issuerAuth: Data

    static func == (lhs: IssuerSigned, rhs: IssuerSigned) -> Bool {
        lhs.nameSpaces == rhs.nameSpaces && lhs.issuerAuth == rhs.issuerAuth
    }
}

/// A single data element within a namespace, wrapped with random salt and digest ID.
struct IssuerSignedItem: Equatable {
    let digestId: Int
    let random: Data
    let elementIdentifier: String
    let elementValue: Any

    static func == (lhs: IssuerSignedItem, rhs: IssuerSignedItem) -> Bool {
        lhs.digestId == rhs.digestId && lhs.elementIdentifier == rhs.elementIdentifier
    }
}

/// Mobile Security Object — the signed payload inside issuerAuth (COSE_Sign1).
struct MobileSecurityObject {
    let version: String
    let digestAlgorithm: String
    let valueDigests: [String: [Int: Data]]
    let deviceKeyInfo: DeviceKeyInfo
    let validityInfo: ValidityInfo
}

/// COSE key identifying the device that holds the mdoc.
struct DeviceKeyInfo: Equatable {
    let deviceKey: Data
}

/// Validity window for the MSO.
struct ValidityInfo: Equatable {
    let signed: String
    let validFrom: String
    let validUntil: String
}
